pkg/cloud/scope/shared.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+)
+
+var (
+	// ErrAZSubnetsNotFound is an error when a availability zone is specified but there are
+	// no matching subnets for that availability zone (a.k.a. fault domain)
+	ErrAZSubnetsNotFound = errors.New("no subnets found for supplied availability zone")
+	// ErrLoggerRequired is an error if a logger isn't specified
+	ErrLoggerRequired = errors.New("logger is required")
+	// ErrNotPlaced is an error if there is no placement determined
+	ErrNotPlaced = errors.New("placement not determined")
+)
+
+type placementInput struct {
+	SpecSubnetIDs           []string
+	SpecAvailabilityZones   []string
+	ParentAvailabilityZones []string
+	ControlplaneSubnets     infrav1.Subnets
+}
+
+type subnetsPlacementStratgey interface {
+	Place(input *placementInput) ([]string, error)
+}
+
+func newDefaultSubnetPlacementStrategy(logger logr.Logger) (subnetsPlacementStratgey, error) {
+	if logger == nil {
+		return nil, ErrLoggerRequired
+	}
+
+	return &defaultSubnetPlacementStrategy{
+		logger: logger,
+	}, nil
+}
+
+// defaultSubnetPlacementStrategy is the default strategy for subnet placement
+type defaultSubnetPlacementStrategy struct {
+	logger logr.Logger
+}
+
+// Place works out the subnet placement based on the following precedence:
+// 1. Explicit definition of subnet IDs in the spec
+// 2. If the spec has Availability Zones then get the subnets for these AZs
+// 3. If the parent resource has Availability Zones then get the subnets for these AZs
+// 4. All the subnets from the control plane are used
+// In Cluster API Availability Zone can also be referred to by the name `Failure Domain`
+func (p *defaultSubnetPlacementStrategy) Place(input *placementInput) ([]string, error) {
+	if len(input.SpecSubnetIDs) > 0 {
+		p.logger.V(2).Info("using subnets from the spec")
+		return input.SpecSubnetIDs, nil
+	}
+
+	if len(input.SpecAvailabilityZones) > 0 {
+		p.logger.V(2).Info("determining subnets to use from the spec availability zones")
+		subnetIDs, err := p.getSubnetsForAZs(input.SpecAvailabilityZones, input.ControlplaneSubnets)
+		if err != nil {
+			return nil, fmt.Errorf("getting subnets for spec azs: %w", err)
```

Comment:
Do we want to return an error if there are no subnets for the specified availability zones, or use the next option? Because this is a breaking change if users had wrong Availability zones as we did not have this check before.

Reply:
How about we log the error and then like you say continue to the next rule?

Reply:
Sounds good to me.

Reply:
I was thinking about this, and I've left the error as is. The logic being if you have explicitly said you want the machine pool in a specific AZ/failure domain and there are no subnets in that az/failure domain then I would see that as an error. What do you think @sedefsavas @randomvariable ?

(like the review point above I have changed the release note to `action required` so that when we do a release users are aware of the change in behaviour)